### PR TITLE
Azure container app deployment

### PR DESCRIPTION
**Pull request by WimVergouwe**
Container app deploy template with support for image tagging and deletion prevention.

Output after 3 new ci/cd runs:
```
Tagging myproject.web:20230607.6 as myproject.web:20230607.6-dev
Disallowing delete of myproject.web:20230607.6-dev
Exluding tags from cleanup: 20230607.5-dev, 20230607.6-dev
Removing tag: myproject.web:20230607.4-dev
```
Documentation: https://wiki.delaware.pro/display/DO/Containers

### CODE REVIEW

File: azure/pipelines/deploy/container-app.yml [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+      $revisions = az containerapp revision list --name ${{ parameters.containerApp }} --resource-group $resourceGroup | ConvertFrom-JSON
+      $currentTags = @( $revisions | % { $_.properties.template.containers } | % { $_.image.Split(':')[-1] } )
+      $currentTags += $pinnedImage.Split(':')[-1]
+
+      # Remove all tags for this environment that are NOT relevant anymore.
+      $allTags = az acr repository show-tags --name $registry --repository $repository | ConvertFrom-JSON
+      $tagsToRemove = $allTags | Where-Object { $_.EndsWith('-${{ parameters.environment }}') } | Where-Object { $_ -notin $currentTags }
+      foreach ($tag in $tagsToRemove) {
+        Write-Host "Removing tag: $($repository):$tag"
+        az acr repository untag --name $registry --image "$($repository):$tag"
```
Comment (dejonghee) [inline azure/pipelines/deploy/container-app.yml:113]: We don't need to do anything with the "--delete-enabled false" we set before (on line 100)?

File: azure/pipelines/deploy/container-app.yml [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+      $pinnedImage = "$image-${{ parameters.environment }}"
+      Write-Host "##vso[task.setvariable variable=pinnedImage]$pinnedImage"
+     
+      Write-Host "Tagging $image as $pinnedImage"
+      az acr import --name $registry --source "$registry/$image" --image $pinnedImage --force
+
+      # Disallow deployed image modification/deletion so that it will not automatically cleaned up by purge tasks.
+      # In other words, lock the image.
+      Write-Host "Disallowing delete of $pinnedImage"
+      az acr repository update --name $registry --image $pinnedImage --write-enabled false --delete-enabled false
```
Comment (dugardeynh) [inline azure/pipelines/deploy/container-app.yml:100]: I expected to have some commands in here to undo this action for images that are no longer relevant. Or is that done by simply removing the tag?

File: azure/pipelines/deploy/container-app.yml (hunk 1 above)
Comment (WimVergouwe) [reply azure/pipelines/deploy/container-app.yml]: Good point

File: azure/pipelines/deploy/container-app.yml (hunk 2 above)
Comment (WimVergouwe) [reply azure/pipelines/deploy/container-app.yml:100]: Right, have added that before removing the tag

File: azure/pipelines/deploy/container-app.yml [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+      # Remove all tags for this environment that are NOT relevant anymore.
+      $allTags = az acr repository show-tags --name $registry --repository $repository | ConvertFrom-JSON
+      $tagsToRemove = $allTags | Where-Object { $_.EndsWith('-${{ parameters.environment }}') } | Where-Object { $_ -notin $currentTags }
+      foreach ($tag in $tagsToRemove) {
+        $imageToRemove = "$($repository):$tag"
+        Write-Host "Removing tag: $imageToRemove"
+
+        # Unlock tag.
+        az acr repository update --name $registry --image $imageToRemove --write-enabled true --delete-enabled true
```
Comment (dugardeynh) [inline azure/pipelines/deploy/container-app.yml:116]: Just to be sure: Are _write-enabled_ and _delete-enabled_ attributes of the tag or the image?

In other words: Image x is/was deployed on dev and qa. It has 3 tags x, x-dev and x-qa.
We want to deploy image y (a new version) to dev. 
Image x is no longer deployed on dev, but is deployed on qa.
Tag x-dev is removed from image x and _write-enabled_ and _delete-enabled_ are reset.
Is tag x-qa still locked or not?

**Review by dejonghee — APPROVED** [review]

File: azure/pipelines/deploy/container-app.yml [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    scriptLocation: 'inlineScript'
+    failOnStandardError: true
+    inlineScript: |
+      $container = Get-Content "${{ parameters.containerJsonPath }}" | ConvertFrom-Json
+      $registry = $($container.registry -replace '.azurecr.io$')
+      $repository = $($container.repository)
+      $image = "$($repository):$($container.version)"
+      
+      # Allow implicit Azure CLI extension install (container app).
+      #az config set extension.use_dynamic_install=yes_without_prompt
```
Comment (WimVergouwe) [inline azure/pipelines/deploy/container-app.yml:89]: Cleanup

File: azure/pipelines/deploy/container-app.yml (hunk 3 above)
Comment (WimVergouwe) [reply azure/pipelines/deploy/container-app.yml]: Yes x-qa is still locked (have tested that manually).
The lock is on the tag.

File: azure/pipelines/deploy/container-app.yml [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+      
+      # Allow implicit Azure CLI extension install (container app).
+      az config set extension.use_dynamic_install=yes_without_prompt
+      
+      # Tag image with a environment specific name so we keep track of the image that is deployed.
+      $pinnedImage = "$image-${{ parameters.environment }}"
+      Write-Host "##vso[task.setvariable variable=pinnedImage]$pinnedImage"
+     
+      Write-Host "Tagging $image as $pinnedImage"
+      az acr import --name $registry --source "$registry/$image" --image $pinnedImage --force
```
[the remark this answers is not in the document]
Comment (WimVergouwe) [reply azure/pipelines/deploy/container-app.yml]: Yes

File: azure/pipelines/deploy/container-app.yml [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+      # Disallow deployed image modification/deletion so that it will not automatically cleaned up by purge tasks.
+      # In other words, lock the image.
+      Write-Host "Disallowing delete of $pinnedImage"
+      az acr repository update --name $registry --image $pinnedImage --write-enabled false --delete-enabled false > $null
+      
+      # Gather all tags that are relevant to the deployed environment.
+      # This means all images/tags that are currently deployed and the one we're about to deploy.
+      # Currently deployed images should be available in environment variable.
+      $currentImages = "$(${{ parameters.variablePrefix }}${{ parameters.module }}RevisionImages)"
```
Comment (dugardeynh) [inline azure/pipelines/deploy/container-app.yml:103]: Who/what is responsible for ensuring this variable exists and it has the expected content? Is it default Azure or a custom implementation?

**Review by dugardeynh — APPROVED** [review]

**Pull request merged**